x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/BaseRunAsSuperuserCommand.java

```diff
@@ -0,0 +1,313 @@
+import org.elasticsearch.xpack.security.support.FileAttributesChecker;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.file.Path;
+import java.security.SecureRandom;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * A {@link KeyStoreAwareCommand} that can be extended fpr any CLI tool that needs to allow a local user with
+ * filesystem write access to perform actions on the node as a superuser. It leverages temporary file realm users
+ * with a `superuser` role.
+ */
+public abstract class BaseRunAsSuperuserCommand extends KeyStoreAwareCommand {
+
+    private static final String[] ROLES = new String[] { "superuser" };
+    private static final int PASSWORD_LENGTH = 14;
+
+    private final OptionSpecBuilder force;
+    private final Function<Environment, CommandLineHttpClient> clientFunction;
+    private final CheckedFunction<Environment, KeyStoreWrapper, Exception> keyStoreFunction;
+    private SecureString password;
+    private String username;
+    final SecureRandom secureRandom = new SecureRandom();
+
+    public BaseRunAsSuperuserCommand(
+        Function<Environment, CommandLineHttpClient> clientFunction,
+        CheckedFunction<Environment, KeyStoreWrapper, Exception> keyStoreFunction,
+        String description
+    ) {
+        super(description);
+        this.clientFunction = clientFunction;
+        this.keyStoreFunction = keyStoreFunction;
+        force = parser.acceptsAll(List.of("f", "force"),
+            "Use this option to force execution of the command against a cluster that is currently unhealthy.");
+    }
+
+    @Override
+    public void close() {
+        if (password != null) {
+            password.close();
+        }
+    }
+
+    @Override
+    protected final void execute(Terminal terminal, OptionSet options, Environment env) throws Exception {
+        validate(terminal, options, env);
+        KeyStoreWrapper keyStoreWrapper = keyStoreFunction.apply(env);
+        final Environment newEnv;
+        final Settings settings;
+        if (keyStoreWrapper != null) {
+            decryptKeyStore(keyStoreWrapper, terminal);
+            Settings.Builder settingsBuilder = Settings.builder();
+            settingsBuilder.put(env.settings(), true);
+            if (settingsBuilder.getSecureSettings() == null) {
+                settingsBuilder.setSecureSettings(keyStoreWrapper);
+            }
+            settings = settingsBuilder.build();
+            newEnv = new Environment(settings, env.configFile());
+        } else {
+            newEnv = env;
+            settings = env.settings();
+        }
+
+        ensureFileRealmEnabled(settings);
```

Comment:
Because file realms don't have Secure Settings, I would move this before the password prompt, and maybe make it the base implementation of validate.

Reply:
We chatted and we're moving `ensureFileRealmEnabled(settings);` higher up but not adding it to a default implementation of `validate()`